File: worker/lib/workers/notification.ex
# Copyright (C) 2021-2025 tiksan
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

defmodule Tornium.Workers.Notification do
  require Logger
  alias Tornium.Repo
  import Ecto.Query

  use Oban.Worker,
    max_attempts: 3,
    priority: 5,
    queue: :notifications,
    tags: ["notification"],
    unique: [
      period: :infinity,
      fields: [:worker, :args],
      keys: [:resource, :resource_id, :restricted],
      states: :incomplete
    ]

  @impl Oban.Worker
  def perform(
        %Oban.Job{
          args: %{
            "notifications" => notifications_ids,
            "resource" => resource,
            "resource_id" => resource_id,
            "restricted" => _restricted?,
            "api_call_id" => api_call_id
          }
        } = _job
      ) do
    api_call_id
    |> Tornium.API.Store.pop()
    |> do_perform(notifications_ids, resource, resource_id)
  end

  @spec do_perform(
          response :: map() | :not_ready | :expired | nil,
          notifications_ids :: [String.t()],
          resource :: String.t(),
          resource_id :: integer()
        ) :: Oban.Worker.result()
  defp do_perform(response, notifications_ids, _resource, _resource_id)
       when is_nil(response) and is_list(notifications_ids) do
    # When a notification is cancelled because of an invalid call ID, we still want to update the next execution to ensure that
    # the notification is run again within an appropriate timeframe.

    notifications =
      Tornium.Schema.Notification
      |> where([n], n.nid in ^notifications_ids)
      |> join(:inner, [n], t in assoc(n, :trigger), on: t.tid == n.trigger_id)
      |> preload([n, t, s], trigger: t)
      |> Repo.all()

    notifications
    |> Enum.uniq_by(fn %Tornium.Schema.Notification{} = notification -> notification.trigger end)
    |> Enum.each(fn %Tornium.Schema.Notification{trigger: %Tornium.Schema.Trigger{} = trigger} ->
      Tornium.Notification.update_next_execution(trigger)
    end)

    {:cancel, :invalid_call_id}
  end

  defp do_perform(:expired = _response, notifications_ids, _resource, _resource_id) when is_list(notifications_ids) do
    # When a notification is cancelled because of an expired API call result, we still want to update the next
    # execution to ensure that the notification is run again within an appropriate timeframe.

    notifications =
      Tornium.Schema.Notification
      |> where([n], n.nid in ^notifications_ids)
      |> join(:inner, [n], t in assoc(n, :trigger), on: t.tid == n.trigger_id)
      |> preload([n, t, s], trigger: t)
      |> Repo.all()

    notifications
    |> Enum.uniq_by(fn %Tornium.Schema.Notification{} = notification -> notification.trigger end)
    |> Enum.each(fn %Tornium.Schema.Notification{trigger: %Tornium.Schema.Trigger{} = trigger} ->
      Tornium.Notification.update_next_execution(trigger)
    end)

    {:cancel, :expired}
  end

  defp do_perform(:not_ready = _response, _notification_ids, _resource, _resource_id) do
    {:error, :not_ready}
  end

  defp do_perform(%{"error" => %{"code" => error_code}} = _response, _notifications_ids, _resource, _resource_id)
       when error_code in [] do
    # This handles error codes that need to be handled in a specific manner. Generic Torn API errors can be handled by
    # cancelling or erroring the job.

    # TODO: Implement this
  end

  defp do_perform(%{"error" => %{"code" => error_code}} = _response, _notifications_ids, _resource, _resource_id)
       when is_integer(error_code) do
    # TODO: Implement this
  end

  defp do_perform(response, notifications_ids, resource, _resource_id)
       when is_map(response) and is_list(notifications_ids) do
    notifications =
      Tornium.Schema.Notification
      |> where([n], n.nid in ^notifications_ids)
      |> join(:inner, [n], t in assoc(n, :trigger), on: t.tid == n.trigger_id)
      |> join(:inner, [n, t], s in assoc(n, :server), on: s.sid == n.server_id)
      |> preload([n, t, s], trigger: t, server: s)
      |> Repo.all()

    notifications
    |> Enum.uniq_by(fn %Tornium.Schema.Notification{} = notification -> notification.trigger end)
    |> Enum.each(fn %Tornium.Schema.Notification{trigger: %Tornium.Schema.Trigger{} = trigger} ->
      Tornium.Notification.update_next_execution(trigger)
    end)

    resource = String.to_atom(resource)

    notifications
    |> Enum.group_by(& &1.trigger_id)
    |> Enum.each(fn {_trigger_id, [%Tornium.Schema.Notification{trigger: trigger} | _] = trigger_notifications} ->
      response
      |> Tornium.Notification.filter_response(resource, trigger.selections)
      |> Tornium.Notification.handle_api_response(trigger, trigger_notifications)
    end)
  end

  @impl Oban.Worker
  def timeout(_job) do
    :timer.seconds(30)
  end
end

File: application/controllers/api/v1/notification/notification.py
# Copyright (C) 2021-2023 tiksan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import json
import uuid

from flask import request
from peewee import DoesNotExist
from tornium_commons.models import Notification

from controllers.api.v1.decorators import ratelimit, session_required
from controllers.api.v1.utils import api_ratelimit_response, make_exception_response


@session_required
@ratelimit
def update_guild_notification(notification_id, *args, **kwargs):
    data = json.loads(request.get_data().decode("utf-8"))
    key = f"tornium:ratelimit:{kwargs['user'].tid}"

    try:
        notification_uuid = uuid.UUID(notification_id)
    except ValueError:
        return make_exception_response("1000", key, details={"message": "Invalid notification UUID"})

    try:
        notification: Notification = (
            Notification.select().where(Notification.nid == notification_uuid).get()
        )  # TODO: Limit selections
    except DoesNotExist:
        return make_exception_response("1401", key)

    if kwargs["user"].tid not in notification.server.admins:
        return make_exception_response("4020", key)

    try:
        channel_id = int(data["channel_id"])
    except (KeyError, ValueError, TypeError):
        return make_exception_response("1001", key)

    try:
        resource_id = int(data["resource_id"])
    except (KeyError, ValueError, TypeError):
        return make_exception_response("1000", key, details={"message": "Invalid resource ID"})

    if resource_id == 0:
        return make_exception_response("1000", key, details={"message": "Invalid resource ID: must not be zero"})

    try:
        one_shot = data["one_shot"]
    except (KeyError, ValueError, TypeError):
        return make_exception_response("1000", key, details={"message": "Invalid one-shot value"})

    if not isinstance(one_shot, bool):
        return make_exception_response("1000", key, details={"message": "Invalid one-shot value"})

    try:
        parameters = data["parameters"]
    except (KeyError, ValueError, TypeError):
        return make_exception_response("1402", key)

    if not isinstance(parameters, dict):
        return make_exception_response("1402", key)
    elif not all(isinstance(key, str) for key in parameters.keys()):
        return make_exception_response("1402", key)
    elif parameters.keys() != notification.trigger.parameters.keys():
        return make_exception_response("1402", key, details={"message": "Invalid parameter key"})

    Notification.update(channel_id=channel_id, resource_id=resource_id, one_shot=one_shot, parameters=parameters).where(
        Notification.nid == notification_uuid
    ).execute()

    return (
        {
            "nid": notification.nid,
        },
        200,
        api_ratelimit_response(key),
    )


@session_required
@ratelimit
def delete_guild_notification(notification_id, *args, **kwargs):
    key = f"tornium:ratelimit:{kwargs['user'].tid}"

    try:
        notification_uuid = uuid.UUID(notification_id)
    except ValueError:
        return make_exception_response("1000", key, details={"message": "Invalid notification UUID"})

    try:
        notification: Notification = (
            Notification.select().where(Notification.nid == notification_uuid).get()
        )  # TODO: Limit selections
    except DoesNotExist:
        return make_exception_response("1401", key)

    if kwargs["user"].tid not in notification.server.admins:
        return make_exception_response("4020", key)

    notification.delete_instance()

    return {}, 200, api_ratelimit_response(key)
